File: sdks/kotlin/src/main/java/com/agentation/mobile/AgentationHttpServer.kt
package com.agentation.mobile

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Lightweight HTTP server that exposes element inspection data on port 4748.
 * The bridge (bridge-android) connects to this via ADB port forwarding:
 *   `adb forward tcp:4748 tcp:4748`
 *
 * Endpoints:
 *   GET /agentation/elements     → JSON array of all tracked MobileElement
 *   GET /agentation/element?x=&y= → hit-test element at coordinates
 *   GET /agentation/health       → { "status": "ok" }
 */
class AgentationHttpServer(
    private val port: Int = 4748,
) {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = false; prettyPrint = false }
    private var serverSocket: ServerSocket? = null
    private var executor: ExecutorService? = null
    @Volatile
    private var running = false

    /**
     * Start the HTTP server on a background thread.
     * Safe to call multiple times — subsequent calls are no-ops.
     */
    fun start() {
        if (running) return
        running = true

        executor = Executors.newCachedThreadPool { r ->
            Thread(r, "agentation-http").apply { isDaemon = true }
        }

        executor?.execute {
            try {
                serverSocket = ServerSocket(port)
                while (running) {
                    try {
                        val socket = serverSocket?.accept() ?: break
                        executor?.execute { handleConnection(socket) }
                    } catch (_: Exception) {
                        if (!running) break
                    }
                }
            } catch (_: Exception) {
                // Port might already be in use — silently fail for dev tool
            }
        }
    }

    /**
     * Stop the HTTP server and release resources.
     */
    fun stop() {
        running = false
        try {
            serverSocket?.close()
        } catch (_: Exception) {}
        executor?.shutdownNow()
        executor = null
        serverSocket = null
    }

    private fun handleConnection(socket: Socket) {
        try {
            socket.soTimeout = 5000
            val reader = BufferedReader(InputStreamReader(socket.getInputStream()))
            val writer = PrintWriter(socket.getOutputStream(), true)

            val requestLine = reader.readLine() ?: return
            // Consume headers
            while (true) {
                val line = reader.readLine() ?: break
                if (line.isEmpty()) break
            }

            val parts = requestLine.split(" ")
            if (parts.size < 2) {
                sendResponse(writer, 400, """{"error":"Bad request"}""")
                return
            }

            val method = parts[0]
            val fullPath = parts[1]
            val pathAndQuery = fullPath.split("?", limit = 2)
            val path = pathAndQuery[0]
            val queryParams = if (pathAndQuery.size > 1) parseQuery(pathAndQuery[1]) else emptyMap()

            if (method != "GET") {
                sendResponse(writer, 405, """{"error":"Method not allowed"}""")
                return
            }

            when (path) {
                "/agentation/health" -> {
                    sendResponse(writer, 200, """{"status":"ok","elements":${ElementInspector.registry.size}}""")
                }
                "/agentation/elements" -> {
                    val body = ElementInspector.toJson()
                    sendResponse(writer, 200, body)
                }
                "/agentation/element" -> {
                    val x = queryParams["x"]?.toDoubleOrNull()
                    val y = queryParams["y"]?.toDoubleOrNull()
                    if (x == null || y == null) {
                        sendResponse(writer, 400, """{"error":"Missing x and y query parameters"}""")
                        return
                    }
                    val element = ElementInspector.elementAt(x, y)
                    if (element != null) {
                        sendResponse(writer, 200, json.encodeToString(element))
                    } else {
                        sendResponse(writer, 404, """{"error":"No element found at coordinates"}""")
                    }
                }
                else -> {
                    sendResponse(writer, 404, """{"error":"Not found"}""")
                }
            }
        } catch (_: Exception) {
            // Silently handle connection errors
        } finally {
            try { socket.close() } catch (_: Exception) {}
        }
    }

    private fun sendResponse(writer: PrintWriter, status: Int, body: String) {
        val statusText = when (status) {
            200 -> "OK"
            400 -> "Bad Request"
            404 -> "Not Found"
            405 -> "Method Not Allowed"
            else -> "Error"
        }
        writer.print("HTTP/1.1 $status $statusText\r\n")
        writer.print("Content-Type: application/json\r\n")
        writer.print("Access-Control-Allow-Origin: *\r\n")
        writer.print("Content-Length: ${body.toByteArray(Charsets.UTF_8).size}\r\n")
        writer.print("Connection: close\r\n")
        writer.print("\r\n")
        writer.print(body)
        writer.flush()
    }

    private fun parseQuery(query: String): Map<String, String> {
        return query.split("&").mapNotNull { param ->
            val kv = param.split("=", limit = 2)
            if (kv.size == 2) kv[0] to java.net.URLDecoder.decode(kv[1], "UTF-8") else null
        }.toMap()
    }
}

File: sdks/swift/Sources/AgentationMobile/AgentationHttpServer.swift
import Foundation

/// Lightweight HTTP server that exposes element inspection data on port 4748.
/// The bridge (bridge-ios) connects to this on the simulator's localhost.
///
/// Endpoints:
///   GET /agentation/elements     → JSON array of all tracked MobileElement
///   GET /agentation/element?x=&y= → hit-test element at coordinates
///   GET /agentation/health       → { "status": "ok" }
///
/// Start from your App's init:
/// ```swift
/// #if DEBUG
/// AgentationHttpServer.shared.start()
/// #endif
/// ```
public final class AgentationHttpServer: @unchecked Sendable {
    public static let shared = AgentationHttpServer()

    private let port: UInt16
    private var listener: Thread?
    private var serverFd: Int32 = -1
    private var running = false

    public init(port: UInt16 = 4748) {
        self.port = port
    }

    /// Start the HTTP server on a background thread.
    public func start() {
        guard !running else { return }
        running = true

        let thread = Thread { [weak self] in
            self?.runServer()
        }
        thread.name = "agentation-http"
        thread.qualityOfService = .utility
        thread.start()
        self.listener = thread
    }

    /// Stop the HTTP server.
    public func stop() {
        running = false
        if serverFd >= 0 {
            close(serverFd)
            serverFd = -1
        }
    }

    private func runServer() {
        serverFd = socket(AF_INET, SOCK_STREAM, 0)
        guard serverFd >= 0 else { return }

        var opt: Int32 = 1
        setsockopt(serverFd, SOL_SOCKET, SO_REUSEADDR, &opt, socklen_t(MemoryLayout<Int32>.size))

        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = port.bigEndian
        addr.sin_addr.s_addr = INADDR_ANY

        let bindResult = withUnsafePointer(to: &addr) { addrPtr in
            addrPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockaddrPtr in
                bind(serverFd, sockaddrPtr, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }

        guard bindResult >= 0 else {
            close(serverFd)
            serverFd = -1
            return
        }

        listen(serverFd, 5)

        while running {
            var clientAddr = sockaddr_in()
            var clientAddrLen = socklen_t(MemoryLayout<sockaddr_in>.size)

            let clientFd = withUnsafeMutablePointer(to: &clientAddr) { addrPtr in
                addrPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockaddrPtr in
                    accept(serverFd, sockaddrPtr, &clientAddrLen)
                }
            }

            guard clientFd >= 0 else {
                if !running { break }
                continue
            }

            DispatchQueue.global(qos: .utility).async { [weak self] in
                self?.handleConnection(clientFd)
            }
        }
    }

    private func handleConnection(_ fd: Int32) {
        defer { close(fd) }

        var buffer = [UInt8](repeating: 0, count: 4096)
        let bytesRead = recv(fd, &buffer, buffer.count, 0)
        guard bytesRead > 0 else { return }

        let request = String(bytes: buffer[0..<bytesRead], encoding: .utf8) ?? ""
        let lines = request.components(separatedBy: "\r\n")
        guard let requestLine = lines.first else { return }

        let parts = requestLine.components(separatedBy: " ")
        guard parts.count >= 2, parts[0] == "GET" else {
            sendResponse(fd, status: 405, body: #"{"error":"Method not allowed"}"#)
            return
        }

        let fullPath = parts[1]
        let pathComponents = fullPath.components(separatedBy: "?")
        let path = pathComponents[0]
        let queryParams = pathComponents.count > 1 ? parseQuery(pathComponents[1]) : [:]

        switch path {
        case "/agentation/health":
            let body = DispatchQueue.main.sync { () -> String in
                let count = ElementInspector.shared.elements.count
                return #"{"status":"ok","elements":\#(count)}"#
            }
            sendResponse(fd, status: 200, body: body)

        case "/agentation/elements":
            let body = DispatchQueue.main.sync { () -> String in
                return ElementInspector.shared.toJson()
            }
            sendResponse(fd, status: 200, body: body)

        case "/agentation/animations":
            let animations = AnimationDetector.shared.getActiveAnimations()
            let encoder = JSONEncoder()
            if let data = try? encoder.encode(animations),
               let json = String(data: data, encoding: .utf8) {
                sendResponse(fd, status: 200, body: json)
            } else {
                sendResponse(fd, status: 200, body: "[]")
            }

        case "/agentation/element":
            guard let xStr = queryParams["x"], let x = Double(xStr),
                  let yStr = queryParams["y"], let y = Double(yStr) else {
                sendResponse(fd, status: 400, body: #"{"error":"Missing x and y query parameters"}"#)
                return
            }
            let result = DispatchQueue.main.sync { () -> (Int, String) in
                if let element = ElementInspector.shared.elementAt(x: x, y: y) {
                    let encoder = JSONEncoder()
                    if let data = try? encoder.encode(element),
                       let json = String(data: data, encoding: .utf8) {
                        return (200, json)
                    }
                    return (500, #"{"error":"Encoding failed"}"#)
                }
                return (404, #"{"error":"No element found at coordinates"}"#)
            }
            sendResponse(fd, status: result.0, body: result.1)

        default:
            sendResponse(fd, status: 404, body: #"{"error":"Not found"}"#)
        }
    }

    private func sendResponse(_ fd: Int32, status: Int, body: String) {
        let statusText: String
        switch status {
        case 200: statusText = "OK"
        case 400: statusText = "Bad Request"
        case 404: statusText = "Not Found"
        case 405: statusText = "Method Not Allowed"
        default: statusText = "Error"
        }

        let bodyData = body.data(using: .utf8) ?? Data()
        let response = "HTTP/1.1 \(status) \(statusText)\r\n" +
            "Content-Type: application/json\r\n" +
            "Access-Control-Allow-Origin: *\r\n" +
            "Content-Length: \(bodyData.count)\r\n" +
            "Connection: close\r\n" +
            "\r\n" +
            body

        if let data = response.data(using: .utf8) {
            data.withUnsafeBytes { ptr in
                if let baseAddress = ptr.baseAddress {
                    send(fd, baseAddress, data.count, 0)
                }
            }
        }
    }

    private func parseQuery(_ query: String) -> [String: String] {
        var result: [String: String] = [:]
        for param in query.components(separatedBy: "&") {
            let kv = param.components(separatedBy: "=")
            if kv.count == 2 {
                result[kv[0]] = kv[1].removingPercentEncoding ?? kv[1]
            }
        }
        return result
    }
}
